### DISCUSSION TITLE
Add support for custom CA.

### DISCUSSION THREAD

**Discussion by rwthompsonii** [k3s-io/k3s#9367] [Q&A]
**Is your feature request related to a problem? Please describe.**

I have a system where I want to build repeatable images with k3s, and give them to downstream developers.  

I want to use certificates signed by a full CA chain that is entirely within my control, to sign into these k3s api servers for auth purposes.

I do not want to check in to source control my entire chain of certs, including the keys.  For either the root or the intermediate.  I do not want private keys involved, or any other secrets.

I just want the ability to specify a trusted CA by the api server, without interfering with k3s's creation of the cert chains that it uses for interoperability internally.

When I give the arg `--k8s-apiserver-arg client-ca-file=/my/root/ca`, it totally ignores the argument and spawns the api server with one in "/etc/rancher/something".  I think that should at least result in a failure to start, not silently drop my argument on the floor.  That would have saved me some debugging, at least.

**Describe the solution you'd like**

I want a flag called "trustThisCA" that will import a root PEM file and trust that root, in addition to any roots that k3s may create.  I do not want to _replace_ k3s root files, I want it to use _both_ when passing them on to the apiserver that k3s is spawning -- I want the convenience of not having to run custom_CA_generate.sh _and_ I want to be able to specify _additional_ root PEMs to trust certs signed by it.

**Describe alternatives you've considered**

N/A
**Additional context**

Most of the input to building a k3s server for me has to be submitted to version control, so putting anything with a secret into it, sucks.  Public keys are great, secrets, not so much.  Being able to say "trust this" without providing a secret is a win, for me.

**Comment by brandond** [comment]
> I do not want to check in to source control my entire chain of certs, including the keys. For either the root or the intermediate. I do not want private keys involved, or any other secrets.

This is not possible. K3s generates a number of certificates for use by internal components. If you want to K3s to sign its internal certs, and validate client certificates, using an existing CA, you MUST provide an intermediate CA certificate+key that K3s can sign additional certificates with.

There is no way around this. If you don't provide the key we can't create the additional required leaf certificates. If the cert is not trusted by your existing CA, then your goal of controlling the root of the CA chain is not met.

Check the docs at https://docs.k3s.io/cli/certificate#custom-ca-topology for more information on supported CA topologies.

Note that what you're doing probably isn't a GREAT idea, as any certificate issued by one cluster will be trusted by all your other clusters, as they are all valid to each other due to the common trusted root CA. Why exactly are you trying to do this?

**Comment by rwthompsonii** [comment]
Sorry, I'll try to make it clearer.

I want:

k3s to make all its own certs, and use those for communication wherever it needs to, and trust _my cert root_ for only k3s api server authentication.  I want an --additional-trusted-root, because I can mint those certs using things outside of k3s, I just need them to be trusted without having to give the keys to the kingdom to it.

**Comment by rwthompsonii** [comment]
As in, I don't _want_ k3s to make leaf certificates from my root.  I want it to use the one it's making up to do that, just like it's doing today.  I just want to be able to authenticate to the API server with certs signed by my root, without mucking with all the rest of that stuff.

**Comment by rwthompsonii** [comment]
Right now I'm going through shenanigans to get a custom intermediate cert minted in CI, and giving it that, but that's kind of a hack that I feel is wildly unnecessary, because any kind of TLS chain should be able to just take "trust the roots in this directory" or something.

**Comment by rwthompsonii** [comment]
As in, I want to take the `client.cert` and `client.key` files I can produce from my own certificate infrastructure, convert them into the format that kubeconfig expects, and then use that kubeconfig to type `kubectl get pods` against k3s, _without_ providing k3s any kind of secret, only the public root PEM that the certs are signed with.

**Reply by brandond** [reply]
That's an interesting use case. Most folks that want to plug an external auth provider into Kubernetes don't want the risk of having an external CA floating around that is capable of issuing trusted certs, since Kubernetes does not support certificate revocation. Instead, most folks use OpenID and point the apiserver at their endpoint with `--kube-apiserver-arg=service-account-issuer=https://your-openid-endpoint/`.

See the upstream docs at https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/ 
> Identifier of the service account token issuer. The issuer will assert this identifier in "iss" claim of issued tokens. This value is a string or URI. If this option is not a valid URI per the OpenID Discovery 1.0 spec, the ServiceAccountIssuerDiscovery feature will remain disabled, even if the feature gate is set to true. It is highly recommended that this value comply with the OpenID spec: https://openid.net/specs/openid-connect-discovery-1_0.html. In practice, this means that service-account-issuer must be an https URL. It is also highly recommended that this URL be capable of serving OpenID discovery documents at {service-account-issuer}/.well-known/openid-configuration. When this flag is specified multiple times, the first is used to generate tokens and all are used to determine which issuers are accepted.

Note that the apiserver supports multiple issuers, so you can add your own while still using the built-in one.

**Reply by rwthompsonii** [reply]
> Note that the apiserver supports multiple issuers, so you can add your own while still using the built-in one.


Can you link me more docs on this one, and I'm looking at OpenID now, tyvm.

**Reply by brandond** [reply]
I linked to the upstream docs.

K3s starts the apiserver with `--service-account-issuer=https://kubernetes.default.svc.cluster.local` which points at the apiserver's built-in issuer, so that will always be the "first" value used to sign tokens. If you add your own issuers via the `--kube-apiserver-arg` flag I described above, those external issuers will be trusted by the apiserver without affecting built-in service account tokens.

This does require the apiserver to be able to reach out to the endpoint in order to retrieve the discovery document, but this is a limitation of how OpenID works.

**Reply by rwthompsonii** [reply]
> That's an interesting use case. Most folks that want to plug an external auth provider into Kubernetes don't want the risk of having an external CA floating around that is capable of issuing trusted certs, since Kubernetes does not support certificate revocation. 

And yes, this is somewhat the reason here, too, in addition to the pain of source control, I don't want to provide a key to something that would allow it to issue certs, even leaf certs, that are signed by my trusted root, because much of the rest of my infrastructure trusts that root unconditionally.  I just want this thing to be able to _accept_ my certs, not mint them.

**Reply by rwthompsonii** [reply]
Oh that's the same doc, didn't realize.  Thanks!

**Reply by brandond** [reply]
There are more docs available at https://kubernetes.io/docs/reference/access-authn-authz/authentication/#configuring-the-api-server

You might try using something basic like https://github.com/CESNET/fake-oidc-server/ just to proof it out. You'd need to call the provider's `authorize` and `token` endpoints to get a bearer token for the `kubernetes` client, and pass it into the kubeconfig in the `token` field of the user for your context:
```yaml
users:
- name: default
  user:
    token: MY-BEARER-TOKEN
```
